### PR TITLE
AirfRANS: vol-weight curriculum 20x→10x decay (front-load volume)

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The static `--vol-loss-weight 10.0` used in the AF champion (PR #3135) applies the same volume penalty throughout training. We hypothesize that **front-loading volume learning early** — starting at 20x and linearly decaying to 10x over the first 50% of epochs — will push the model to learn volumetric structure faster, then let the surface loss dominate refinement in the second half.

This is a curriculum learning approach: high vol-weight early prevents the model from locking into a surface-only attractor; decay prevents the vol-weight from suppressing surface accuracy at convergence.

Target: **vol_mse < 0.0017** with **surface_mse ≤ 0.000296** (no regression).

## Instructions

Implement a dynamic volume weight schedule in `target/icml2026/train.py`. The formula is:

```python
vol_weight(epoch) = start_weight - (start_weight - end_weight) * min(1.0, epoch / decay_epochs)
```

Where:
- `start_weight` = trial-specific starting value (see trials below)
- `end_weight` = 10.0 (the champion static value)
- `decay_epochs` = 50% of total training epochs (i.e. `int(0.5 * args.epochs)`)
- `epoch` is 0-indexed

Add `--vol-weight-start` and `--vol-weight-end` CLI flags (floats). If `--vol-weight-start` is not provided (or equals `--vol-loss-weight`), fall back to the existing static behaviour.

**Log `vol_weight` to W&B each epoch** (e.g. `wandb.log({"vol_weight": current_vol_weight}, step=epoch)`).

**Smoke test first** (3 epochs) to confirm vol_weight logs correctly and loss is finite before launching full runs.

### Trial 1 — Aggressive curriculum (20x → 10x)

Start at 20x, decay to champion 10x over first 50% of epochs. Use the full champion config:

```bash
cd target/icml2026 && python train.py \
  --dataset airfrans --airfrans-task full \
  --optimizer adamw --lr 6e-4 \
  --cosine-t-max 50 --grad-clip 1.0 --weight-decay 1e-2 \
  --enable-fourier \
  --model-layers 2 --model-hidden-dim 256 --model-heads 4 \
  --epochs 999 \
  --ema-decay 0.999 \
  --vol-weight-start 20.0 --vol-weight-end 10.0 \
  --wandb_group af-volume-weight-curriculum
```

### Trial 2 — Gentle curriculum (15x → 10x)

Same config, gentler starting weight:

```bash
cd target/icml2026 && python train.py \
  --dataset airfrans --airfrans-task full \
  --optimizer adamw --lr 6e-4 \
  --cosine-t-max 50 --grad-clip 1.0 --weight-decay 1e-2 \
  --enable-fourier \
  --model-layers 2 --model-hidden-dim 256 --model-heads 4 \
  --epochs 999 \
  --ema-decay 0.999 \
  --vol-weight-start 15.0 --vol-weight-end 10.0 \
  --wandb_group af-volume-weight-curriculum
```

**EMA is critical** — keep `--ema-decay 0.999`. The upweighted volume gradient early in training causes instability without EMA smoothing.

Report both trials in the PR. Primary metric: `vol_mse` (lower is better). Hard constraint: `surface_mse ≤ 0.000296`.

## Baseline

Current AF champion: PR #3135 (nezuko — EMA=0.999 + vol-weight=10x static)

| Metric | Value | Epoch |
|---|---|---|
| surface_mse (val) | 0.000296 | 704 |
| vol_mse (val) | 0.002039 | 743 |

- Baseline W&B run: `sh2zyfwr`
- Reproduce: `cd target/icml2026 && python train.py --dataset airfrans --airfrans-task full --optimizer adamw --lr 6e-4 --cosine-t-max 50 --grad-clip 1.0 --weight-decay 1e-2 --enable-fourier --model-layers 2 --model-hidden-dim 256 --model-heads 4 --epochs 999 --ema-decay 0.999 --vol-loss-weight 10.0`

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT chihiro:

## Results

### Trial 1 — Aggressive curriculum 20x→10x

| Metric | Baseline (PR #3135) | Trial 1 (best val) | Trial 1 (test @ best) | SpiderSolver ref |
|--------|--------------------|--------------------|----------------------|-----------------|
| surface_mse | 0.000296 (ep704) | 0.001295 (ep306) | 0.001422 | 0.0043 |
| volume_mse | 0.002039 (ep743) | 0.005036 (ep299) | 0.005984 | 0.0017 |

- **Epochs completed:** 777 / 999 (360-min timeout)
- **Best checkpoint:** epoch 306 (by val surface_mse)
- **Vol_weight schedule:** 20.0 → 10.0 linear decay over first 499 epochs
- **W&B run:** `75lpm3lz` — [view](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/75lpm3lz)
- **Peak grad_norm_mean:** 200–500+ (severe, vs ~25–35 for stable training)
- **Verdict:** ❌ Clear failure. 4.4x worse surface_mse, 2.5x worse vol_mse vs baseline.

**Instability event (ep335–400):** The 20x starting weight caused severe gradient instability. Surface_mse spiked from ~0.002 to 0.006+, vol_mse from 0.005 to 0.010+. The EMA model preserved the ep306 best checkpoint, but the online model never recovered. Grad norms remained 200–500+ throughout training — 10x higher than Trial 2.

### Trial 2 — Gentle curriculum 15x→10x

| Metric | Baseline (PR #3135) | Trial 2 (best val) | Trial 2 (test @ best) | SpiderSolver ref |
|--------|--------------------|--------------------|----------------------|-----------------|
| surface_mse | 0.000296 (ep704) | 0.000408 (ep595) | 0.000713 | 0.0043 |
| volume_mse | 0.002039 (ep743) | 0.002443 (ep759) | 0.004157 | 0.0017 |

- **Epochs completed:** 765 / 999 (360-min timeout)
- **Best checkpoint:** epoch 595 (by val surface_mse)
- **Vol_weight schedule:** 15.0 → 10.0 linear decay over first 499 epochs
- **W&B run:** `72xdp0o9` — [view](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/72xdp0o9)
- **Peak grad_norm_mean:** 25–35 (stable, normal range)
- **Verdict:** ❌ Better than Trial 1 but still regresses vs baseline. 1.38x worse surface_mse, 1.20x worse vol_mse.

### Commands used

**Trial 1:**
```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=0 python train.py \
  --dataset airfrans --airfrans-task full \
  --optimizer adamw --lr 6e-4 \
  --cosine-t-max 50 --grad-clip 1.0 --weight-decay 1e-2 \
  --enable-fourier \
  --model-layers 2 --model-hidden-dim 256 --model-heads 4 \
  --epochs 999 --ema-decay 0.999 \
  --vol-weight-start 20.0 --vol-weight-end 10.0 \
  --agent chihiro --wandb-name "chihiro/af-vol-curriculum-20x-10x" \
  --wandb-group af-volume-weight-curriculum
```

**Trial 2:**
```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=1 python train.py \
  --dataset airfrans --airfrans-task full \
  --optimizer adamw --lr 6e-4 \
  --cosine-t-max 50 --grad-clip 1.0 --weight-decay 1e-2 \
  --enable-fourier \
  --model-layers 2 --model-hidden-dim 256 --model-heads 4 \
  --epochs 999 --ema-decay 0.999 \
  --vol-weight-start 15.0 --vol-weight-end 10.0 \
  --agent chihiro --wandb-name "chihiro/af-vol-curriculum-15x-10x" \
  --wandb-group af-volume-weight-curriculum
```

### What happened

**The hypothesis is rejected.** Front-loading volume learning via a decaying curriculum does not improve on the static 10x vol-weight baseline. Both trials regressed on both metrics.

Key findings:

1. **20x starting weight is too aggressive.** It causes severe gradient instability that EMA=0.999 cannot fully absorb. The 20x→10x curriculum essentially wastes the first ~300 epochs fighting instability, and the model never catches up.

2. **15x→10x is stable but doesn't help.** The gentle curriculum trains smoothly (grad norms comparable to baseline) but converges to worse metrics than static 10x. The extra 5x volume emphasis in early epochs doesn't create a better initialization for the refinement phase — it just slightly delays surface learning.

3. **The static 10x baseline is already near-optimal for this tradeoff.** The champion's approach (constant 10x + EMA=0.999) appears to give the model the right balance throughout training. A curriculum that starts higher doesn't "front-load" volume learning — it just destabilizes or slows surface convergence without a compensating volume benefit.

4. **Why no volume benefit?** The curriculum decays to 10x by epoch 499, giving the model 266+ epochs at the same weight as baseline. But by that point, baseline has already been at 10x for all 499 epochs and has a head start on both metrics. The curriculum's early high-weight phase hurts surface learning more than it helps volume learning.

### Suggested follow-ups

- **Inverse curriculum (low→high):** Start at 5x or 7x, ramp UP to 10x. Let surface learning lock in first, then increase volume pressure. This is the opposite of what was tried here and might work better given that surface accuracy appears to be the harder target.
- **Shorter decay window:** Instead of 50% of epochs, try 10–20% with a milder starting weight (12x→10x). This would minimize disruption while still testing the curriculum concept.
- **Cyclical vol-weight:** Oscillate between 8x and 12x on a cosine schedule (like cosine LR), which might prevent the model from settling into a suboptimal surface-volume tradeoff.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #3226 because ADVISOR: Closing — vol-weight curriculum is dead for AF. Trial 1 (20x→10x): collapsed ep335, surface 4.4x worse. Trial 2 (15x→10x): stable but 1.38x worse on both metrics. The static 10x value is validated as the optimal operating point — deviating above it (even temporarily) only costs surface convergence without any compensating volume benefit. Good analysis on the gradient norm dynamics at 20x..
